File: frontend/api/locations.ts
import type { LocationItem, LocationResponse } from '~/types/location'

export const useLocationsApi = () => {
  const { $api } = useNuxtApp()
  const countriesLoading = ref(false)
  const federalDistrictsLoading = ref(false)
  const regionsLoading = ref(false)
  const citiesLoading = ref(false)
  const countriesError = ref<Error | null>(null)
  const federalDistrictsError = ref<Error | null>(null)
  const regionsError = ref<Error | null>(null)
  const citiesError = ref<Error | null>(null)

  const countryOptions = ref<LocationItem[]>([])
  const federalDistrictOptions = ref<LocationItem[]>([])
  const regionOptions = ref<LocationItem[]>([])
  const cityOptions = ref<LocationItem[]>([])

  // Загрузка списка стран
  const loadCountries = async () => {
    countriesLoading.value = true
    countriesError.value = null
    try {
      const response = await $api.get('/v1/locations/countries')
      const data = response as LocationResponse
      countryOptions.value = data.items || []
    } catch (error) {
      countriesError.value = error as Error
      console.error('Error loading countries:', error)
    } finally {
      countriesLoading.value = false
    }
  }

  // Загрузка федеральных округов для России
  const loadFederalDistricts = async () => {
    federalDistrictsLoading.value = true
    federalDistrictsError.value = null
    try {
      const response = await $api.get('/v1/locations/federal-districts')
      const data = response as LocationResponse
      federalDistrictOptions.value = data.items || []
    } catch (error) {
      federalDistrictsError.value = error as Error
      console.error('Error loading federal districts:', error)
    } finally {
      federalDistrictsLoading.value = false
    }
  }

  // Загрузка регионов для выбранной страны
  const loadRegions = async (countryCode: string, federalDistrict?: string) => {
    regionsLoading.value = true
    regionsError.value = null
    try {
      const response = await $api.get(`/v1/locations/regions/${countryCode}`, {
        params: { federal_district: federalDistrict }
      })
      const data = response as LocationResponse
      regionOptions.value = data.items || []
    } catch (error) {
      regionsError.value = error as Error
      console.error('Error loading regions:', error)
    } finally {
      regionsLoading.value = false
    }
  }

  // Загрузка городов для выбранного региона
  const loadCities = async (countryCode: string, regionId: string, level?: number, searchQuery?: string) => {
    citiesLoading.value = true
    citiesError.value = null
    try {
      const response = await $api.get(`/v1/locations/cities`, {
        params: { 
          level, 
          region: regionId, 
          country: countryCode,
          name: searchQuery
        }
      })
      const data = response as LocationResponse
      cityOptions.value = data.items || []
    } catch (error) {
      citiesError.value = error as Error
      console.error('Error loading cities:', error)
    } finally {
      citiesLoading.value = false
    }
  }

  // Поиск городов по названию
  const searchCities = async (cityName: string) => {
    citiesLoading.value = true
    citiesError.value = null
    try {
      const response = await $api.get(`/v1/locations/cities/search/${encodeURIComponent(cityName)}`)
      const data = response as LocationResponse
      cityOptions.value = data.items || []
    } catch (error) {
      citiesError.value = error as Error
      console.error('Error searching cities:', error)
    } finally {
      citiesLoading.value = false
    }
  }

  // Загрузка начальных данных
  onMounted(() => {
    loadCountries()
  })

  return {
    // Состояния загрузки
    countriesLoading,
    federalDistrictsLoading,
    regionsLoading,
    citiesLoading,
    
    // Ошибки
    countriesError,
    federalDistrictsError,
    regionsError,
    citiesError,
    
    // Данные
    countryOptions,
    federalDistrictOptions,
    regionOptions,
    cityOptions,
    
    // Методы
    loadCountries,
    loadFederalDistricts,
    loadRegions,
    loadCities,
    searchCities
  }
} 

File: backend/app/api/common/routes/locations.py
from typing import Optional

from fastapi import APIRouter, HTTPException

from app.api.common.schemas.location import (
    LocationResponse,
    CitySearchResponse
)
from app.api.common.utils.location_api import LocationAPI

router = APIRouter(prefix="/locations", tags=["locations"])
location_api = LocationAPI()


@router.get("/countries", response_model=LocationResponse)
async def get_countries():
    """
    Получить список стран.
    """
    try:
        countries = await location_api.get_countries()
        return LocationResponse(
            items=countries,
            total=len(countries)
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/regions/{country_code}", response_model=LocationResponse)
async def get_regions(country_code: str, federal_district: Optional[str] = Query(None)):
    """
    Получить список регионов для указанной страны.
    
    Args:
        country_code: Код страны (например, 'RU')
        federal_district: Название федерального округа (необязательно)
    """
    try:
        regions = await location_api.get_regions(country_code, federal_district)
        return LocationResponse(
            items=regions,
            total=len(regions)
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/cities/{region_id}", response_model=LocationResponse)
async def get_cities(region_id: int, level: Optional[int] = None):
    """
    Получить список городов для указанного региона.
    """
    try:
        cities = await location_api.get_cities(region_id, level)
        return LocationResponse(
            items=cities,
            total=len(cities)
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/cities/search/{city_name}", response_model=CitySearchResponse)
async def search_cities(city_name: str):
    """
    Поиск городов по частичному совпадению названия.
    """
    try:
        cities = await location_api.search_cities(city_name)
        return CitySearchResponse(
            items=cities,
            total=len(cities)
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/districts/{region_id}", response_model=LocationResponse)
async def get_districts(region_id: int):
    """
    Получить список районов для указанного региона.
    """
    try:
        districts = await location_api.get_districts(region_id)
        return LocationResponse(
            items=districts,
            total=len(districts)
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
